Add `no_std` compatibility

haerdib opened the pull request:
- Adds `no_std` compatibility  by adding a `std` feature and changing stds-import to `core` or `alloc`. 
- Updates `scale-info` from v2.2 to v 2.7
- Fixes some clippy deprecated warnings.

haerdib committed:
make no_std compatible

haerdib committed:
fix clippy

haerdib committed:
fix buil

haerdib committed:
fix no-std build

haerdib committed:
fix scale-info feature

haerdib committed:
let serde stay optional

haerdib committed:
fix serde no-std

niklasad1 reviewed (approved):
nice but I noticed that it needs a few more fixes to compile.

haerdib committed:
fix cargo check

haerdib commented:
Fixed it :)

haerdib committed:
fix serde

jsdw commented on src/scale/format.rs:
```diff
@@ -16,6 +16,7 @@
 //! Ths module defines a [`Format`], which is basically a [`StoreFormat`] and an
 //! [`OrderFormat`] and describes the different possible wire formats of a bit sequence.
 
+use alloc::string::String;
```
Nit; missing newline:

```suggestion
use alloc::string::String;

```

haerdib committed:
Update src/scale/format.rs

Co-authored-by: James Wilson <james@jsdw.me>

jsdw commented on Cargo.toml:
```diff
@@ -24,9 +28,9 @@ serde = ["dep:serde"]
-scale-info = { version = "2.2.0", default-features = false, features = ["std"], optional = true }
-codec = { package = "parity-scale-codec", version = "3.2.0", default-features = false, features = [] }
-serde = { version = "1", optional = true }
+scale-info = { version = "2.2.0", default-features = false, optional = true }
```
Let's bump this to 2.5.0 or above, since you're using the pub fields now which requires at least that version :)

jsdw commented on Cargo.toml:
```diff
@@ -14,7 +14,11 @@ keywords = ["parity", "scale", "encoding", "decoding", "bits"]
-default = ["serde", "scale-info"]
+default = ["std", "serde", "scale-info"]
+std = [
+    "scale-info/std",
```
Can we only enable scale-info if it's explicitly requested, or was there a reason to not put the `?` on it here? 

```suggestion
    "scale-info?/std",
```

haerdib replied:
No, I missed that one. Thanks!

haerdib committed:
Update Cargo.toml

Co-authored-by: James Wilson <james@jsdw.me>

haerdib committed:
bump to scale-info v2.7.0

jsdw commented:
Looks awesome! We should also add a CI check to ensure no_std compat isn't accidentally broken going forwards; adding this to the rust.yml file I think will do the trick (this same approach can be used in scale-encode and scale-decode):

```yaml
  no_std:
    name: Check no_std build
    runs-on: ubuntu-latest
    steps:
      - name: Checkout sources
        uses: actions/checkout@v2

      - name: Install Rust stable toolchain
        uses: actions-rs/toolchain@v1
        with:
          profile: minimal
          toolchain: stable
          target: thumbv6m-none-eabi
          override: true

      - name: Rust Cache
        uses: Swatinem/rust-cache@v1.3.0

      - name: Check no_std build
        uses: actions-rs/cargo@v1.0.3
        with:
          command: check
          # The thumbv6m-none-eabi doesn't support `std`, so 
          # will fail if the crate is not no_std compatible.
          args: --no-default-features --target thumbv6m-none-eabi
```

haerdib committed:
add no-std check to CI

haerdib commented:
Added the CI `no_std` check. In addition, I added the features `scale-info` and `serde` to it, that way these features are also checked for no_std compatibility.

jsdw commented:
```
error[E0432]: unresolved import `alloc::sync`
  --> /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/scale-info-2.7.0/src/prelude.rs:50:13
   |
50 |             sync,
   |             ^^^^ no `sync` in the root
   |
   = help: consider importing this module instead:
           core::sync
```

Interesting; I wouldn't expect `scale-info` to have a no-std related issue.

For alloc::sync the docs say:

```
Note: This module is only available on platforms that support atomic loads and stores of pointers. This may be detected at compile time using #[cfg(target_has_atomic = "ptr")].
```

So maybe the no_std target I suggested has this limitation too. Maybe there's a better target choice?

jsdw commented:
The target `aarch64-unknown-none` seems to work offhand, and also only supports `no_std`. (With that, there is one small fix needed to add a `.to_string()` in `scale/format.rs`. because `scale-info` gives &strs and not Strings back when no-std, so the test is being useful :)).

Not sure if there's a reason against picking some targets over others here..

haerdib commented:
I suppose it's fine to start with an architecture that's simple. If there's the need for a specific one that does not work, then there's always the possibility to open an issue. 

Thank you very much for investigating btw!

haerdib committed:
change no_std toolchain to aarch64

jsdw commented:
Ah just need to also change the `--target` in the CI to `aarch64-unknown-none` too (maybe it defaults to the toolchain listed above already; not sure so I played it safe here!)

jsdw committed:
update --target in CI too

jsdw commented:
Edit: I did that :)

jsdw reviewed (approved):
Amazing; thanks for your contribution @haerdib!

I'll get a release of this published shortly.
